### PR TITLE
Add partial-volume-aware label resampling mode

### PR DESCRIPTION
**Pull request by fepegar**
[Partially generated by a coding agent]

---

image

image

image

---

## Summary

Adds an opt-in `label_interpolation="label"` mode to the spatial transforms (`Spatial`, `Resample`, `Affine`, `ElasticDeformation`) for **proper, partial-volume-aware resampling of label maps**.

Nearest-neighbor interpolation (the default) produces staircase artifacts, biases label volumes, and erodes small structures under strong downsampling. The new mode instead:

1. one-hot encodes the label map using the **unique** label values present (robust to non-contiguous labels such as `{0, 2, 5}`);
2. optionally Gaussian-smooths each channel before downsampling when `antialias=True` (reusing the existing Cardoso et al., MICCAI 2015 sigma);
3. resamples every channel with **linear** interpolation; and
4. takes the per-voxel `argmax`, remapping back to the original label values.

Out-of-bounds voxels use `default_pad_label`. Multi-channel (already one-hot) label inputs are resampled linearly without re-encoding.

## Behavior / compatibility

- **Defaults unchanged**: `label_interpolation` stays `"nearest"` everywhere; this is purely opt-in.
- `image_interpolation="label"` raises a clear `ValueError` (the mode is label-only).
- The mode never produces label values absent from the input.

## Example

```python
import torchio as tio

transform = tio.Resample(2, label_interpolation="label", antialias=True)
```

## Tests

New `TestLabelInterpolation` in `tests/test_spatial.py`: no invalid labels (up/downsampling), non-contiguous round-trip, `default_pad_label` fill, `image_interpolation="label"` raises, antialias path, multi-channel input, and a round-trip Dice check showing `"label"` beats `"nearest"`.

`uv run pytest tests/test_spatial.py tests/test_one_hot.py -q` -> 99 passed. `ruff` and `ty` clean.

## Notes

- For *extreme* downsampling of highly convoluted multi-tissue maps, nearest can occasionally match/beat `"label"` on Dice because `argmax` correctly drops sub-voxel-thin structures; `"label"` wins clearly at moderate-to-strong factors and on small structures.
- A separate, pre-existing bug was observed during testing: `copy.deepcopy` of a `CropOrPad` output reverts the image to its original (pre-crop) shape, which makes `Resample` compute the wrong output grid (since `forward` deepcopies inputs). Not addressed here; can file separately.

### CODE REVIEW

File: src/torchio/transforms/spatial/spatial.py [hunk 1]
```diff
@@ -199,11 +206,22 @@ class Spatial(SpatialTransform):
...
+            `"label"`.  Used for [`LabelMap`][torchio.LabelMap] instances.
+            The `"label"` mode performs partial-volume-aware resampling:
+            the label map is one-hot encoded, each channel is resampled
+            with linear interpolation, and the per-voxel argmax recovers
+            the discrete labels.  Compared with `"nearest"`, this reduces
+            staircase artifacts and yields more accurate label volumes,
+            which is especially useful when downsampling.  It never
+            produces label values that were absent from the input, and is
+            more memory- and compute-intensive because it processes one
+            channel per label.
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/spatial/spatial.py]: [Generated by a coding agent]

Fixed in 2a45bcdb. Reworded the `label_interpolation` docstring to note that the only new value that can appear is `default_pad_label`, used for out-of-bounds voxels.

File: src/torchio/transforms/spatial/spatial.py [hunk 2]
```diff
@@ -714,26 +744,123 @@ def _apply_spatial_to_batch(
...
+       (robust to non-contiguous labels such as $\{0, 2, 5\}$),
+    2. optionally Gaussian-smooths each channel before downsampling when
+       *antialias* is `True` (using the same sigma as
+       [`_antialias_sigmas`][] from Cardoso et al., MICCAI 2015),
+    3. resamples every channel with linear interpolation, and
+    4. takes the per-voxel argmax to recover the discrete labels.
+
+    Out-of-bounds voxels are filled with *default_pad_label*.
+
+    Args:
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/spatial/spatial.py]: [Generated by a coding agent]

Fixed in 2a45bcdb. The helper docstring now states explicitly that the single-channel path fills out-of-bounds with `default_pad_label`, while the multi-channel path fills with 0 and does not apply `default_pad_label`.

File: src/torchio/transforms/spatial/spatial.py [hunk 3]
```diff
@@ -714,26 +744,123 @@ def _apply_spatial_to_batch(
...
+            probabilistic map) the channels are resampled linearly without
+            re-encoding.
+        grid: `(I_out, J_out, K_out, 3)` sampling grid in input voxel
+            coordinates.
+        input_shape: Spatial shape of the input volume `(I, J, K)`.
+        input_affine: Affine of the input grid.
+        output_affine: Affine of the output grid.
+        antialias: Whether to Gaussian-smooth the one-hot channels before
+            downsampling.
+        default_pad_label: Value assigned to out-of-bounds voxels.
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/spatial/spatial.py]: [Generated by a coding agent]

Fixed in 2a45bcdb. The `default_pad_label` parameter doc now states it is only applied for single-channel (`C == 1`) inputs; multi-channel inputs use 0 for out-of-bounds.

File: src/torchio/transforms/spatial/spatial.py [hunk 4]
```diff
@@ -714,26 +744,123 @@ def _apply_spatial_to_batch(
...
+        smoothed = data.float()
+        if antialias:
+            smoothed = _antialias_batch(smoothed, input_affine, output_affine)
+        return _sample_batch(
+            smoothed,
+            grid,
+            input_shape=input_shape,
+            interpolation="linear",
+            fill_value=0.0,
+        ).to(data.dtype)
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/spatial/spatial.py]: [Generated by a coding agent]

Fixed in 2a45bcdb. The multi-channel branch now preserves floating-point output for non-floating inputs (e.g. integer one-hot encodings), so partial volumes are no longer truncated to 0/1. Added test `test_multichannel_integer_input_preserves_partial_volumes`.

File: src/torchio/transforms/spatial/spatial.py [hunk 5]
```diff
@@ -714,24 +746,136 @@ def _apply_spatial_to_batch(
...
+    Nearest-neighbor resampling of a label map produces staircase artifacts
+    and can drop thin structures, especially when downsampling. This helper
+    instead:
+
+    1. one-hot encodes the label map using the unique label values present
+       (robust to non-contiguous labels such as $\{0, 2, 5\}$),
+    2. optionally Gaussian-smooths each channel before downsampling when
+       *antialias* is `True` (using the same sigma as
+       [`_antialias_sigmas`][] from Cardoso et al., MICCAI 2015),
+    3. resamples every channel with linear interpolation, and
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/spatial/spatial.py]: [Generated by a coding agent]

Fixed in 1a54dd40. Replaced the `[`_antialias_sigmas`][]` shortcut reference (a private helper not exposed in the docs) with plain inline code.

File: tests/test_spatial.py [hunk 6]
```diff
@@ -827,3 +827,112 @@ def test_order_0_uses_fast_path(self) -> None:
...
+        subject = tio.Subject(seg=tio.LabelMap(original, affine=np.eye(4)))
+
+        def roundtrip(mode: str) -> torch.Tensor:
+            down = tio.Resample(4, label_interpolation=mode)(subject)
+            back = tio.Resample(subject.seg, label_interpolation=mode)(down)
+            return back.seg.data
+
+        dice_label = _dice(roundtrip("label"), original)
+        dice_nearest = _dice(roundtrip("nearest"), original)
+        assert dice_label > dice_nearest
```
[the remark this answers is not in the document]
Comment (fepegar) [reply tests/test_spatial.py]: [Generated by a coding agent]

Fixed in 1a54dd40. Relaxed the assertion to `dice_label >= dice_nearest` (not worse than nearest) with a comment, to avoid flakiness from grid alignment or library-version ties.

File: src/torchio/transforms/spatial/spatial.py [hunk 7]
```diff
@@ -91,10 +91,16 @@
     "fifth",
     "sixth",
     "seventh",
+    "label",
 ]
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/spatial/spatial.py]: [Generated by a coding agent]

Fixed in 5dccd885. Split the alias into `TypeImageInterpolation` (without `"label"`) and `TypeLabelInterpolation = TypeImageInterpolation | Literal["label"]`. `image_interpolation` parameters and attributes now use the narrower type, so `image_interpolation="label"` is now rejected by static type checkers as well as at runtime. `TypeInterpolation` is kept as a broad alias for internal helpers.

File: src/torchio/transforms/spatial/spatial.py [hunk 8]
```diff
@@ -714,24 +754,136 @@ def _apply_spatial_to_batch(
...
+        interpolation="linear",
+        fill_value=0.0,
+    )
+
+    winners = sampled.argmax(dim=1)
+    resampled = labels[winners]
+    # In-bounds voxels keep partition of unity (channels sum to ~1); voxels
+    # sampled entirely from outside the input have all-zero channels.
+    in_bounds = sampled.sum(dim=1) > 0.5
+    resampled = torch.where(
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/spatial/spatial.py]: [Generated by a coding agent]

The 0.5 threshold is intentional: it matches the existing fill convention for intensity images in `_sample_batch_grid_sample` (`torch.where(mask > 0.5, sampled, fill_tensor)` at line ~1225), so label and intensity padding behave the same at borders (a voxel sampled mostly, >50%, from outside is treated as out-of-bounds). The misleading part was the comment, which claimed only all-zero/entirely-outside voxels were affected. Updated the comment in a6120b8f to describe the actual >50%-outside behavior and reference the shared convention.

File: src/torchio/transforms/spatial/spatial.py [hunk 9]
```diff
@@ -199,11 +210,24 @@ class Spatial(SpatialTransform):
             transforms.
         image_interpolation: `"linear"` (default) or `"nearest"`.
             Used for [`ScalarImage`][torchio.ScalarImage] instances.
-        label_interpolation: `"nearest"` (default) or `"linear"`.
-            Used for [`LabelMap`][torchio.LabelMap] instances.
+        label_interpolation: `"nearest"` (default), `"linear"`, or
+            `"label"`.  Used for [`LabelMap`][torchio.LabelMap] instances.
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/spatial/spatial.py]: [Generated by a coding agent]

Fixed in a6120b8f. The docstrings now state that both `image_interpolation` and `label_interpolation` also accept the higher-order B-spline modes (`"quadratic"`..`"seventh"`) and the equivalent integer orders 0-7.

File: src/torchio/transforms/spatial/spatial.py [hunk 10]
```diff
@@ -199,11 +210,24 @@ class Spatial(SpatialTransform):
...
+            The `"label"` mode performs partial-volume-aware resampling:
+            the label map is one-hot encoded, each channel is resampled
+            with linear interpolation, and the per-voxel argmax recovers
+            the discrete labels.  Compared with `"nearest"`, this reduces
+            staircase artifacts and yields more accurate label volumes,
+            which is especially useful when downsampling.  It never
+            invents intermediate label values that were absent from the
+            input (the only new value that can appear is
+            `default_pad_label`, used for out-of-bounds voxels), and is
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/spatial/spatial.py]: [Generated by a coding agent]

Fixed in a6120b8f. The `label_interpolation` docstring now distinguishes the two paths: single-channel inputs never invent intermediate labels, while multi-channel (one-hot/probabilistic) maps are resampled linearly per channel and can contain fractional partial volumes.

**Comment by romainVala** [comment]
Hi there

This is a cool add on !
 I use a 'home made' version of this interpolation, which is indeed very useful especially when you want to upsample a label file

Just a small comment about step 3
`3 .resamples every channel with linear interpolation; `

 It would be nice to add an extra parametre (let call it one_hot_label_interpolation=linear by default ) but the user could then also use other interpolation (bspline, sinc ect ...)

Cheers 

Romain

(may be a note to add in the documentation:  this can be extremely memory demanding as the number of labels augments)
I work now with more thant 300 labels, and this is a typical use case where applying this transform patch by patch with a gridSampler would help to adjust this process to available memory

**Comment by fepegar** [comment]
Thanks, @romainVala. Your suggestions have been addressed in e7f8e0f2b5f9b209a79a91493c1a3e73c8991c83.
